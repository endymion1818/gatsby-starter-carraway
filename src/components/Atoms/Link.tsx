import GatsbyLink from 'gatsby-link'
import React, { FC } from 'react'
import styled, { css } from 'styled-components'
import * as token from '../tokens'
import ButtonStyles from './ButtonStyles'

export interface ILinkProps {
  openInNewTab?: boolean
  to?: string
  href?: string
  rel?: string
  noUnderline?: boolean
  activeClassName?: string
  isButton?: boolean
  onClick?: (args: unknown) => void
}

const LinkStyles = css<ILinkProps>`
  ${({ noUnderline }) => (noUnderline ? `text-decoration: none` : `text-decoration: underline;`)}
  text-decoration-skip-ink: auto;

  &:focus {
    box-shadow: 0 0 2px ${token.colors.base.primary};
  }
  ${({ isButton }) => isButton && ButtonStyles}
`

const SLink = styled(GatsbyLink)<ILinkProps>`
  ${LinkStyles}
`

const Anchor = styled.a<ILinkProps>`
  ${LinkStyles}
`

const Link: FC<ILinkProps> = ({ children, to = '', openInNewTab, noUnderline, ...other }) => {
  // This regex assumes that any internal link (intended for Gatsby to process)
  // will start with exactly one slash, and that anything else is external.
  const internal = /^\/(?!\/)/.test(to)

  // This regex checks for file extensions - any string including a dot followed by
  // a series of other charachters.

  const file = /\.[0-9a-z]+$/i.test(to)

  // Use gatsby-link for internal links, <a> for others, or <div> if no to or href is specified.
  // If internal, use the `file` regex to determine if this is a file resource, permitting a download.

  if (typeof to === 'undefined' && typeof other.href === 'undefined') {
    return <span {...other}>{children}</span>
  }

  if (internal) {
    if (file) {
      return (
        <Anchor href={to} {...other} noUnderline={noUnderline}>
          {children}
        </Anchor>
      )
    }
    return (
      <SLink to={to} {...other} noUnderline={noUnderline}>
        {children}
      </SLink>
    )
  }
  return (
    <>
      {openInNewTab ? (
        <Anchor href={to} {...other} target="_blank" rel="noopener" noUnderline={noUnderline}>
          {children}
        </Anchor>
      ) : (
        <Anchor href={to} {...other} noUnderline={noUnderline}>
          {children}
        </Anchor>
      )}
    </>
  )
}
export default Link
