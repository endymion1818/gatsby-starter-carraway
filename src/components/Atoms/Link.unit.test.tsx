import { cleanup, render } from '@testing-library/react'
import React from 'react'
import Link from './Link'

describe('<Link />', () => {
  beforeEach(cleanup)
  it('Should render an <a> tag', () => {
    const { container } = render(<Link href="https://deliciousreverie.co.uk">test</Link>)
    expect(container).toMatchSnapshot()
  })
  it('Should render a <span>', () => {
    const { container } = render(<Link>test</Link>)
    expect(container).toMatchSnapshot()
  })
})
