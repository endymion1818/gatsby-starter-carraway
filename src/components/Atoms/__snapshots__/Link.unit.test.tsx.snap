// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> Should render a <span> 1`] = `
.c0 {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-text-decoration-skip-ink: auto;
  text-decoration-skip-ink: auto;
}

.c0:focus {
  box-shadow: 0 0 2px #000080;
}

<div>
  <a
    class="c0"
    href=""
  >
    test
  </a>
</div>
`;

exports[`<Link /> Should render an <a> tag 1`] = `
.c0 {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-text-decoration-skip-ink: auto;
  text-decoration-skip-ink: auto;
}

.c0:focus {
  box-shadow: 0 0 2px #000080;
}

<div>
  <a
    class="c0"
    href="https://deliciousreverie.co.uk"
  >
    test
  </a>
</div>
`;
